# Importamos librerias

In [13]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np
import os


import sys 
sys.path.append("..")
from src import fx_limpieza as fx_l


# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

## Importamos todos los CSV y los nombramos con su respectivo año

In [14]:
df_13 = pd.read_csv("../datos/raw_data/datos-2013.csv", sep=';', encoding='latin-1')
df_14 = pd.read_csv("../datos/raw_data/datos-2014.csv", sep=';', encoding='latin-1')
df_15 = pd.read_csv("../datos/raw_data/datos-2015.csv", sep=';', encoding='latin-1')
df_16 = pd.read_csv("../datos/raw_data/datos-2016.csv", sep=';', encoding='latin-1')
df_17 = pd.read_csv("../datos/raw_data/datos-2017.csv", sep=';', encoding='latin-1')
df_18 = pd.read_csv("../datos/raw_data/datos-2018.csv", sep=';', encoding='latin-1')
df_19 = pd.read_csv("../datos/raw_data/datos-2019.csv", sep=';', encoding='latin-1')
df_20 = pd.read_csv("../datos/raw_data/datos-2020.csv", sep=';', encoding='latin-1')
df_21 = pd.read_csv("../datos/raw_data/datos-2021.csv", sep=';', encoding='latin-1')
diccionario = pd.read_csv("../datos/raw_data/diccionario_datos.csv", sep=';', encoding='latin-1')

## Concatenamos todos los Dataframes

In [15]:
df = pd.concat([df_13, df_14, df_15,df_16, df_17, df_18, df_19, df_20, df_21], axis=0, ignore_index = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                     Non-Null Count    Dtype 
---  ------                     --------------    ----- 
 0   CÓDIGO ÓRGÃO SUPERIOR      1026299 non-null  int64 
 1   NOME ÓRGÃO SUPERIOR        1026299 non-null  object
 2   CÓDIGO ÓRGÃO               1026299 non-null  int64 
 3   NOME ÓRGÃO                 1026299 non-null  object
 4   CÓDIGO UNIDADE GESTORA     1026299 non-null  int64 
 5   NOME UNIDADE GESTORA       1026299 non-null  object
 6   CATEGORIA ECONÔMICA        1026299 non-null  object
 7   ORIGEM RECEITA             1026299 non-null  object
 8   ESPÉCIE RECEITA            1026299 non-null  object
 9   DETALHAMENTO               1026299 non-null  object
 10  VALOR PREVISTO ATUALIZADO  1026299 non-null  object
 11  VALOR LANÇADO              1026299 non-null  object
 12  VALOR REALIZADO            1026299 non-null  object
 13  PERCENTUAL REALIZADO       

## Renombramos columnas para reducir barrera de lenguaje

In [16]:
df.rename(columns={
    "CÓDIGO ÓRGÃO SUPERIOR": "COD_organo_superior",
    "NOME ÓRGÃO SUPERIOR": "NOM_organo_superior",
    "CÓDIGO ÓRGÃO": "COD_organo",
    "NOME ÓRGÃO": "NOM_organo",
    "CÓDIGO UNIDADE GESTORA": "COD_unidad_gestora",
    "NOME UNIDADE GESTORA": "NOM_unidad_gestora",
    "CATEGORIA ECONÔMICA": "categoria_economica",
    "ORIGEM RECEITA": "origen_ingreso",
    "ESPÉCIE RECEITA": "especie_ingreso",
    "DETALHAMENTO": "detalle_adicional ",
    "VALOR PREVISTO ATUALIZADO": "valor_previsto_actualizado",
    "VALOR LANÇADO": "valor_lanzado",
    "VALOR REALIZADO": "valor_recaudado",
    "PERCENTUAL REALIZADO": "porcentaje_ejecutado",
    "DATA LANÇAMENTO": "fecha_ejecucion",
    "ANO EXERCÍCIO": "anio_ejecucion"
}, inplace=True)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                      Non-Null Count    Dtype 
---  ------                      --------------    ----- 
 0   COD_organo_superior         1026299 non-null  int64 
 1   NOM_organo_superior         1026299 non-null  object
 2   COD_organo                  1026299 non-null  int64 
 3   NOM_organo                  1026299 non-null  object
 4   COD_unidad_gestora          1026299 non-null  int64 
 5   NOM_unidad_gestora          1026299 non-null  object
 6   categoria_economica         1026299 non-null  object
 7   origen_ingreso              1026299 non-null  object
 8   especie_ingreso             1026299 non-null  object
 9   detalle_adicional           1026299 non-null  object
 10  valor_previsto_actualizado  1026299 non-null  object
 11  valor_lanzado               1026299 non-null  object
 12  valor_recaudado             1026299 non-null  object
 13  porcentaje_e

## Pasamos todos los valores que tenemos como objeto a formato numero

In [18]:
df["valor_previsto_actualizado"] = df["valor_previsto_actualizado"].str.replace(",", ".").astype(float)

In [19]:
df["valor_lanzado"] = df["valor_lanzado"].str.replace(",", ".").astype(float)

In [20]:
df["valor_recaudado"] = df["valor_recaudado"].str.replace(",", ".").astype(float)

In [21]:
df["porcentaje_ejecutado"] = df["porcentaje_ejecutado"].str.replace(",", ".").astype(float)

## Pasamos los valores que tenemos como objeto a formato fecha


In [22]:
# Convertimos la columna a formato datetime
df["fecha_ejecucion"] = pd.to_datetime(df["fecha_ejecucion"], errors="coerce", dayfirst="True")

## Chequeamos los valores nulos por columnas

In [23]:
## Llamamos a la funcion de nuestra biblioteca de funciones.
fx_l.valores_nulos(df)

numero_nulos  porcentaje_nulos       tipo_dato
COD_organo_superior                    0              0.00           int64
NOM_organo_superior                    0              0.00          object
COD_organo                             0              0.00           int64
NOM_organo                             0              0.00          object
COD_unidad_gestora                     0              0.00           int64
NOM_unidad_gestora                     0              0.00          object
categoria_economica                    0              0.00          object
origen_ingreso                         0              0.00          object
especie_ingreso                        0              0.00          object
detalle_adicional                      0              0.00          object
valor_previsto_actualizado             0              0.00         float64
valor_lanzado                          0              0.00         float64
valor_recaudado                        0              0.00         float64
porcentaje_ejecutado                   0              0.00         float64
fecha_ejecucion                      578              0.06  datetime64[ns]
anio_ejecucion                         0              0.00           int64

#### Solamente tenemos nulos en la columna de fecha de ejecucion. Lo que se puede hacer es reemplazarlos con el año que tenemos en la columna de al lado y ponerle 01-01-año

#### Igualmente vemos que muchas columnas relacionadas con los valores o float tienen como valor 0

In [24]:
df.sample(10)

COD_organo_superior                            NOM_organo_superior  \
79933                 25000                         Ministério da Economia   
253121                25000                         Ministério da Economia   
292251                26000                         Ministério da Educação   
867233                36000                            Ministério da Saúde   
989570                26000                         Ministério da Educação   
724494                41000                    Ministério das Comunicações   
202419                54000                          Ministério do Turismo   
666457                26000                         Ministério da Educação   
789715                25000                         Ministério da Economia   
573838                22000  Ministério da Agricultura, Pecuária e Abastec   

        COD_organo                                         NOM_organo  \
79933        25000  Ministério da Economia - Unidades com vínculo ...   
253121       37904        Fundo do Regime Geral da Previdência Social   
292251       26255  Universidade Federal dos Vales do Jequitinhonh...   
867233       36201                              Fundação Oswaldo Cruz   
989570       26414                   Instituto Federal do Mato Grosso   
724494       41903  Fundo para o Desenvolvimento Tecnológico das T...   
202419       34902                          Fundo Nacional de Cultura   
666457       26419             Instituto Federal do Rio Grande do Sul   
789715       25000  Ministério da Economia - Unidades com vínculo ...   
573838       22211                Companhia Nacional de Abastecimento   

        COD_unidad_gestora                             NOM_unidad_gestora  \
79933               170013        SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
253121              513001   COORDENACAO DE ORCAMENTO E FINANCAS DO FRGPS   
292251              153036  UNIVERSIDADE FED.VALES JEQUITINHONHA E MUCURI   
867233              254420                          FUNDACAO OSWALDO CRUZ   
989570              158144   INST.FED.DE EDUC.,CIENC.E TEC.DO MATO GROSSO   
724494              410007   FUNDO P/O DESENV.TECNOL.DAS TELECOMUNICACOES   
202419              340002    COORD.GERAL DE PLAN. ORC. FIN. E CONTAB/FNC   
666457              158141            INST.FED.DE EDUC.,CIENC.E TEC.DO RS   
789715              170013        SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
573838              135100            COMPANHIA NACIONAL DE ABASTECIMENTO   

       categoria_economica                               origen_ingreso  \
79933   Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
253121  Receitas Correntes                    Outras Receitas Correntes   
292251  Receitas Correntes                          Receita de Serviços   
867233  Receitas Correntes                          Receita Patrimonial   
989570  Receitas Correntes                    Outras Receitas Correntes   
724494  Receitas Correntes                                Contribuições   
202419  Receitas Correntes                    Outras Receitas Correntes   
666457  Receitas Correntes                          Receita de Serviços   
789715  Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
573838  Receitas Correntes                    Outras Receitas Correntes   

                                      especie_ingreso  \
79933                                        Impostos   
253121                      Demais receitas correntes   
292251   Serviços Administrativos e Comerciais Gerais   
867233  Exploração do patrimônio imobiliário do Estad   
989570    Indenizações, restituições e ressarcimentos   
724494                       Contribuições econômicas   
202419    Indenizações, restituições e ressarcimentos   
666457   Serviços Administrativos e Comerciais Gerais   
789715                                       Impostos   
573838    Indenizações, restituições e ressarcimentos   

                                   detalle_

In [25]:
# Reemplaza los nulos y convierte a fecha
df["fecha_ejecucion"] = df["fecha_ejecucion"].fillna(
    df["anio_ejecucion"].astype(str)+"-01-01"  # Supone que el formato es YYYY-MM-DD
)


In [27]:
#Comprobamos que hayan desaparecido los nulos
print("Valores nulos por columna:")
display(df.isna().sum())

Valores nulos por columna:


COD_organo_superior           0
NOM_organo_superior           0
COD_organo                    0
NOM_organo                    0
COD_unidad_gestora            0
NOM_unidad_gestora            0
categoria_economica           0
origen_ingreso                0
especie_ingreso               0
detalle_adicional             0
valor_previsto_actualizado    0
valor_lanzado                 0
valor_recaudado               0
porcentaje_ejecutado          0
fecha_ejecucion               0
anio_ejecucion                0
dtype: int64

## Tratamiento de duplicados

In [28]:
duplicados = df[df.duplicated()]
duplicados

Empty DataFrame
Columns: [COD_organo_superior, NOM_organo_superior, COD_organo, NOM_organo, COD_unidad_gestora, NOM_unidad_gestora, categoria_economica, origen_ingreso, especie_ingreso, detalle_adicional , valor_previsto_actualizado, valor_lanzado, valor_recaudado, porcentaje_ejecutado, fecha_ejecucion, anio_ejecucion]
Index: []

#### No se encontraron valores duplicados en nuestro dataframe

**Insights:** 

- El DataFrame contiene 1026299 registros y 16 columnas con información sobre características de recaudacion y ejecucion de presupuestos de diferentes ministerios

- Despues de la limpieza de datos la columnas tienen datos completos 

- Los tipos de datos incluyen 4 columnas numéricas enteras (int64), 7 categóricas (object), 4 numérica flotante (float64) y 1 de fecha (datetime64[ns]). 

- El DataFrame ocupa aproximadamente 125.3+ KB en memoria.

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                      Non-Null Count    Dtype         
---  ------                      --------------    -----         
 0   COD_organo_superior         1026299 non-null  int64         
 1   NOM_organo_superior         1026299 non-null  object        
 2   COD_organo                  1026299 non-null  int64         
 3   NOM_organo                  1026299 non-null  object        
 4   COD_unidad_gestora          1026299 non-null  int64         
 5   NOM_unidad_gestora          1026299 non-null  object        
 6   categoria_economica         1026299 non-null  object        
 7   origen_ingreso              1026299 non-null  object        
 8   especie_ingreso             1026299 non-null  object        
 9   detalle_adicional           1026299 non-null  object        
 10  valor_previsto_actualizado  1026299 non-null  float64       
 11  valor_lanzado           

# Finalizada la limpieza de datos, exportamos el df limpio

In [30]:
df.to_pickle("../datos/output_data/df_limpio.pkl")